databuilder/query_engines/sqlite.py

```diff
@@ -168,6 +168,25 @@ def get_sql_month_from_date(self, node):
     def get_sql_day_from_date(self, node):
         return self.get_date_part(self.get_sql(node.source), "DAY")
 
+    @get_sql.register(Function.DateDifferenceInYears)
+    def get_sql_date_difference_in_years(self, node):
+        return self.date_difference_in_years(
+            self.get_sql(node.lhs), self.get_sql(node.rhs)
+        )
+
+    def date_difference_in_years(self, start, end):
+        year_d = self.get_date_part(end, "YEAR") - self.get_date_part(start, "YEAR")
+        month_d = self.get_date_part(end, "MONTH") - self.get_date_part(start, "MONTH")
+        day_d = self.get_date_part(end, "DAY") - self.get_date_part(start, "DAY")
+        # The CASE condition below is a simplification of:
+        #
+        #    end_month < start_month OR (end_month == start_month AND end_day < start_day)
+        #
+        # which doesn't involve multiple references to the "get month" function. The
+        # coefficient here is arbitrary as long as it's at least 31 — we use a power of
+        # two for neatness.
```

Comment:
This is quite a thing. Is it a standard trick for comparing dates?

I had to do quite a lot of work to convince myself that it's correct, so I think a more detailed explanation is in order. I'll try to draft something based on my (newly acquired) understanding of what's going on. I'm assuming changes to address my two naming comments above, but these three suggestions are all orthogonal.

I also think that using 32 as the multiplier is misleading because it's hard to shake the belief that it being `31 + 1` is necessary. I'd prefer that we either use 31 or some much larger number. I'll fold an assumption for changing this into my proposal too, but again this is orthogonal.

> The CASE condition below detects the situation where the end day-of-year is earlier than the start day-of-year. In such cases we need to subtract one from the year delta to give the number of whole years that has elapsed between the two dates.
> 
> The most obvious way to determine this duplicates calls to the "get month" function:
>
>     end_month < start_month OR (end_month == start_month AND end_day < start_day)
>
> or equivalently:
> 
>     delta_month < 0 OR (delta_month == 0 AND delta_day < 0)
>
> However this expression has a simplified form that takes advantage of the fact that `delta_day` has a maximum magnitude of `30`:
> 
>     delta_month * -31 > delta_day
>
> To satisfy yourself that these expressions are equivalent, consider this truth table:
>
>      delta_month | delta_day | result
>     =============|===========|========
>          +ve     |    any    | false
>     -------------|-----------|--------
>                  |    +ve    | false
>                  |-----------|--------
>           0      |     0     | false
>                  |-----------|--------
>                  |    -ve    | true
>     -------------|-----------|--------
>          -ve     |    any    | true
>
> By inspection, this truth table is correct for both expressions and therefore they are equivalent.